*  Copyright 2025 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description : Handle DNS resolver to send query to DNS server with link local address.
* Version : v001
* Mandatory/Optional : Mandatory
* Impact if not applied : thread 1.4 BR_1 DNS 11.3 fails.
*/

Upstream-Status: Backport

commit c020f862301f31c83bbdfbe27031a8f4e4f4bd53
Author: Yang Song <yangsongcn@google.com>
Date:   Tue Jun 3 02:09:27 2025 +0800

    [posix] DNS resolver to handle link-local server address (#11548)
    
    When `sendto` is used with a Link-Local IPv6 destination, the index of
    the outgoing network interface must be included.

Upstream-Status: Inappropriate  [platform specific changes]

diff --git a/src/posix/platform/resolver.cpp b/src/posix/platform/resolver.cpp
index ffd74ec51..a9fcbe10d 100644
--- a/src/posix/platform/resolver.cpp
+++ b/src/posix/platform/resolver.cpp
@@ -231,8 +231,8 @@ otError Resolver::SendQueryToServer(Transaction        *aTxn,
 {
     otError      error = OT_ERROR_NONE;
     otIp4Address ip4Addr;
-    sockaddr_in  serverAddr4;
-    sockaddr_in6 serverAddr6;
+    sockaddr_in  serverAddr4 = {};
+    sockaddr_in6 serverAddr6 = {};
 
     if (otIp4FromIp4MappedIp6Address(&aServerAddress, &ip4Addr) == OT_ERROR_NONE)
     {
@@ -249,6 +249,11 @@ otError Resolver::SendQueryToServer(Transaction        *aTxn,
         memcpy(&serverAddr6.sin6_addr, &aServerAddress, sizeof(otIp6Address));
         serverAddr6.sin6_family = AF_INET6;
         serverAddr6.sin6_port   = htons(53);
+        if (IsIp6AddressLinkLocal(aServerAddress))
+        {
+            // Network interface index is required for link local destinations
+            serverAddr6.sin6_scope_id = otSysGetInfraNetifIndex();
+        }
 
         VerifyOrExit(sendto(aTxn->mUdpFd6, aPacket, aLength, MSG_DONTWAIT, reinterpret_cast<sockaddr *>(&serverAddr6),
                             sizeof(serverAddr6)) > 0,
